server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -548,19 +548,35 @@ final SearchContext createAndPutContext(ShardSearchRequest request) throws IOExc
         }
 
         SearchContext context = createContext(request);
+        onNewContext(context);
         boolean success = false;
         try {
             putContext(context);
-            if (request.scroll() != null) {
+            // ensure that if index is deleted concurrently, we free the context immediately, either here or in freeAllContextForIndex
+            indicesService.indexServiceSafe(request.shardId().getIndex());
```

Comment:
I am still not sure why we do this here. Why can't we just let the search go and fail whenever?

Reply:
It might be overkill... The way I understand the code, the primary purpose of `afterIndexRemoved` is to ensure that we free memory and file system resources as quickly as possible when an index is deleted/closed. Here we try to ensure the same happens also during a race condition. Without this, we theoretically risk holding on to memory/file system resources for a while after the index is deleted/closed. With this, we should free such resources as soon as the index is deleted (and any current search phases have run to completion).

I guess there is a trade-off in that doing this double validation has a small cost and the likelihood of seeing the race condition is very small. I can certainly remove this double validation if you think?

Reply:
I think we should remove the double validation. But the fix you added in this PR is crucial!

Reply:
Thanks @s1monw , I have removed the double validation, please have another look at your convenience.